client/network/src/service.rs

```diff
@@ -1237,11 +1439,82 @@ impl<B: BlockT + 'static, H: ExHashT> Future for NetworkWorker<B, H> {
 							.inc();
 					}
 				},
-				Poll::Ready(SwarmEvent::Behaviour(BehaviourOut::Event(ev))) => {
+				Poll::Ready(SwarmEvent::Behaviour(BehaviourOut::NotificationStreamOpened { remote, engine_id, notifications_sink, role })) => {
 					if let Some(metrics) = this.metrics.as_ref() {
-						metrics.update_with_network_event(&ev);
+						metrics.notifications_streams_opened_total
+							.with_label_values(&[&maybe_utf8_bytes_to_string(&engine_id)]).inc();
+					}
+					{
+						let mut peers_notifications_sinks = this.peers_notifications_sinks.lock();
+						peers_notifications_sinks.insert((remote.clone(), engine_id), notifications_sink);
 					}
-					this.event_streams.send(ev);
+					this.event_streams.send(Event::NotificationStreamOpened {
+						remote,
+						engine_id,
+						role,
+					});
+				},
+				Poll::Ready(SwarmEvent::Behaviour(BehaviourOut::NotificationStreamReplaced { remote, engine_id, notifications_sink })) => {
+					let mut peers_notifications_sinks = this.peers_notifications_sinks.lock();
+					if let Some(s) = peers_notifications_sinks.get_mut(&(remote, engine_id)) {
+						*s = notifications_sink;
+					} else {
+						log::error!(
+							target: "sub-libp2p",
+							"NotificationStreamReplaced for non-existing substream"
+						);
+					}
+
+					// TODO: Notifications might have been lost as a result of the previous
+					// connection being dropped, and as a result it would be preferable to notify
+					// the users of this fact by simulating the substream being closed then
+					// reopened.
```

Comment:
At the moment, unless sending a notification has returned an error, there is a guarantee that the notification will eventually reach its destination.
A `NotificationStreamClosed` indeeds means that a certain number of notifications have potentially not been received.

> The only way I see that client code could actually react to knowing that "some notifications may not have been sent" is if it would always buffer the last n notifications it sent, for some n, just in case it gets a NotificationStreamClosed followed by a NotificationStreamOpened for the same protocol, at which point it could start by sending out all the notifications in the buffer again

Users are not expected to buffer up notifications in case a `NotificationStreamClosed` arrives.
However users are expected to reach to `NotificationStreamOpened` by sending out some message.

---

Let's take the example of the "neighbour" message of the GRANDPA protocol. I'm not familiar with the details, but if I'm not mistaken this message indicates what is the current GRANDPA round of the local node.

Whenever the local node switches to a new GRANDPA round, it indicates that by sending out a "neighbour" message to all its peers. When a `NotificationStreamOpened` event is received, it also sends out a "neighbour" message to the peer we have just connected to.

In the situation where a connection closes and we switch to a different one, by generating a `NotificationStreamClosed` followed by `NotificationStreamOpened`, we would force GRANDPA to send out this "neighbour" message that has potentially been lost.

While this is a specific example, simulating a disconnect/reconnect forces the local node to refresh all the state that it shares with this peer.

Reply:
> At the moment, unless sending a notification has returned an error, there is a guarantee that the notification will eventually reach its destination.

I wouldn't know what that is supposed to tell me as a client of the API, however - i.e. what consequences I can take away for my code. As a client who just sends notifications and is neither aware of the sizes of local buffers nor of remote buffers never truly knows, just from sending notifications without getting an error, whether a notification has been received and processed. It can only reason along the lines of "I sent this notification so-and-so long ago and there was no error since, so it is very likely that the remote has received and processed this notification".

But I think I understand what you're getting at. I think the actual guarantee in question here is "__if__ the remote received notification Y and I sent notification X before Y on the same notification protocol without error, then it follows that the remote received notification X as well.", i.e. that there are no "silent" gaps, which can certainly occur if `sc-network` transparently switches to a different connections or substreams for sending notifications of a certain protocol when the first connection or substream gets closed or errors, without reporting that closure or error on the API. But even this guarantee, since it is a conditional, only ever truly means something for the sender if it ever learns that a notification has actually been received and processed, e.g. by means of a response or indirectly through inbound notifications from the same peer.

It's just that, in practice, any requirements on ordered and gap-less delivery of a notifications protocol usually transcend the life of a single connection and thus having an ordered, reliable transport connection is not enough to make the protocol solid in that sense. Even when just using a single connection, it eventually breaks down. A new connection needs to be established on which you want to continue sending notifications, ideally starting from the last notification you know the remote received. However, here at the boundaries between an old connection and a new one, there are again no guarantees about the ordering in which notifications arrive that have been sent just before the old connection died and those sent just after the new connection opened, since these are sent on different connections. Similarly, without first asking the remote about its current state w.r.t. the notifications received, the sender doesn't even know how many, if any, of the notifications sent shortly before the old connection broke down made it to the receiver. Hence to make the protocol resilient, it still needs to have a way to deal with gaps and out-of-order delivery in received notifications one way or another (the latter may be considered "temporary gap"s, I guess). So an ordered and reliable transport greatly assists in achieving ordered and reliable protocols, but it can usually not provide the full story for an application-level protocol whose messages have a lifetime and ordering beyond that of a single connection.

Essentially it seems to me an interesting question about what level of abstraction (and associated guarantees) `sc-network` wants to provide on the API for notification protocols: The analogue of a transport-level connection/substream or an abstraction over connections/substreams for "fire-and-forget" protocols that most of the time provide reliable and ordered delivery but may occasionally fail to do so as a result of e.g. switching connections internally. I just tend to think that for client code and most such protocols it practically makes not much of a difference in the sense that if you want reliable and ordered delivery of all the protocol messages to some peer, then the guarantees of a single connection or substream don't suffice either way.

> Whenever the local node switches to a new GRANDPA round, it indicates that by sending out a "neighbour" message to all its peers. When a NotificationStreamOpened event is received, it also sends out a "neighbour" message to the peer we have just connected to. Thanks to this, we are guaranteed that all of our peers always know our current local GRANDPA round.

As you say (and I would expect), such notifications are sent out regularly and so missing one should not be an issue. Allowing for the occasional missed message and / or using auxiliary pull-based protocols to fill gaps are really important for protocols that don't want to be unnecessarily fragile. In any case, I have no objection to leaving the TODO in here, of course. This is an interesting discussion to have.